// @aws-amplify/ui-react-liveness/dist/esm/components/FaceLivenessDetector/service/utils/createStreamingClient/createStreamingClient.mjs
import { StartFaceLivenessSessionCommand, RekognitionStreamingClient } from '@aws-sdk/client-rekognitionstreaming';
import { getAmplifyUserAgent } from '@aws-amplify/core/internals/utils';
import { isString } from '@aws-amplify/ui';
import { getLivenessUserAgent } from '../../../utils/platform.mjs';
import { CONNECTION_TIMEOUT, queryParameterString } from '../constants.mjs';
import { CustomWebSocketFetchHandler } from './CustomWebSocketFetchHandler.mjs';
import { resolveCredentials } from './resolveCredentials.mjs';
import { Signer } from './Signer.mjs';
import { createTelemetryReporterMiddleware } from '../TelemetryReporter/TelemetryReporter.mjs';

const CUSTOM_USER_AGENT = `${getAmplifyUserAgent()} ${getLivenessUserAgent()}`;
async function getStreamingClient({ credentialsProvider, endpointOverride, region, systemClockOffset, }) {
    const clientconfig = {
        credentials: await resolveCredentials(credentialsProvider),
        customUserAgent: CUSTOM_USER_AGENT,
        region,
        requestHandler: new CustomWebSocketFetchHandler({
            connectionTimeout: CONNECTION_TIMEOUT,
        }),
        signerConstructor: Signer,
        systemClockOffset,
    };
    if (isString(endpointOverride)) {
        clientconfig.endpointProvider = () => ({ url: new URL(endpointOverride) });
    }
    return new RekognitionStreamingClient(clientconfig);
}
const createCommandInput = ({ requestStream, sessionId, videoWidth, videoHeight, }) => ({
    ChallengeVersions: queryParameterString,
    SessionId: sessionId,
    LivenessRequestStream: requestStream,
    VideoWidth: videoWidth,
    VideoHeight: videoHeight,
});
/**
 * Initializes an instance of the Rekognition streaming client, returns `getResponseStream`
 *
 * @async
 * @param clientConfig configuration fpr the client
 * @returns {Promise<{ getResponseStream: GetReponseStream }>}
 */
async function createStreamingClient(clientConfig) {
    const client = await getStreamingClient(clientConfig);
    client.middlewareStack.add(createTelemetryReporterMiddleware(clientConfig.attemptCount, clientConfig.preCheckViewEnabled), {
        step: 'build',
        name: 'telemetryMiddleware',
        tags: ['liveness', 'amplify-ui'],
    });
    return {
        async getResponseStream(input) {
            const command = new StartFaceLivenessSessionCommand(createCommandInput(input));
            const { LivenessResponseStream } = await client.send(command);
            return LivenessResponseStream;
        },
    };
}

export { createStreamingClient };
