// @aws-sdk/client-rekognitionstreaming/dist-es/commands/StartFaceLivenessSessionCommand.js
import { getEventStreamPlugin } from "@aws-sdk/middleware-eventstream";
import { getWebSocketPlugin } from "@aws-sdk/middleware-websocket";
import { getEndpointPlugin } from "@smithy/middleware-endpoint";
import { Command as $Command } from "@smithy/smithy-client";
import { commonParams } from "../endpoint/EndpointParameters";
import { StartFaceLivenessSession$ } from "../schemas/schemas_0";
export { $Command };
export class StartFaceLivenessSessionCommand extends $Command
    .classBuilder()
    .ep(commonParams)
    .m(function (Command, cs, config, o) {
    return [
        getEndpointPlugin(config, Command.getEndpointParameterInstructions()),
        getEventStreamPlugin(config),
        getWebSocketPlugin(config, {
            headerPrefix: 'x-amz-rekognition-streaming-liveness-',
        }),
    ];
})
    .s("RekognitionStreamingService", "StartFaceLivenessSession", {
    eventStream: {
        input: true,
        output: true,
    },
})
    .n("RekognitionStreamingClient", "StartFaceLivenessSessionCommand")
    .sc(StartFaceLivenessSession$)
    .build() {
}
