// @aws-sdk/client-rekognitionstreaming/dist-cjs/index.js
'use strict';

var middlewareEventstream = require('@aws-sdk/middleware-eventstream');
var middlewareHostHeader = require('@aws-sdk/middleware-host-header');
var middlewareLogger = require('@aws-sdk/middleware-logger');
var middlewareRecursionDetection = require('@aws-sdk/middleware-recursion-detection');
var middlewareUserAgent = require('@aws-sdk/middleware-user-agent');
var middlewareWebsocket = require('@aws-sdk/middleware-websocket');
var configResolver = require('@smithy/config-resolver');
var core = require('@smithy/core');
var schema = require('@smithy/core/schema');
var eventstreamSerdeConfigResolver = require('@smithy/eventstream-serde-config-resolver');
var middlewareContentLength = require('@smithy/middleware-content-length');
var middlewareEndpoint = require('@smithy/middleware-endpoint');
var middlewareRetry = require('@smithy/middleware-retry');
var smithyClient = require('@smithy/smithy-client');
var httpAuthSchemeProvider = require('./auth/httpAuthSchemeProvider');
var runtimeConfig = require('./runtimeConfig');
var regionConfigResolver = require('@aws-sdk/region-config-resolver');
var protocolHttp = require('@smithy/protocol-http');

const resolveClientEndpointParameters = (options) => {
    return Object.assign(options, {
        useDualstackEndpoint: options.useDualstackEndpoint ?? false,
        useFipsEndpoint: options.useFipsEndpoint ?? false,
        defaultSigningName: "rekognition",
    });
};
const commonParams = {
    UseFIPS: { type: "builtInParams", name: "useFipsEndpoint" },
    Endpoint: { type: "builtInParams", name: "endpoint" },
    Region: { type: "builtInParams", name: "region" },
    UseDualStack: { type: "builtInParams", name: "useDualstackEndpoint" },
};

const getHttpAuthExtensionConfiguration = (runtimeConfig) => {
    const _httpAuthSchemes = runtimeConfig.httpAuthSchemes;
    let _httpAuthSchemeProvider = runtimeConfig.httpAuthSchemeProvider;
    let _credentials = runtimeConfig.credentials;
    return {
        setHttpAuthScheme(httpAuthScheme) {
            const index = _httpAuthSchemes.findIndex((scheme) => scheme.schemeId === httpAuthScheme.schemeId);
            if (index === -1) {
                _httpAuthSchemes.push(httpAuthScheme);
            }
            else {
                _httpAuthSchemes.splice(index, 1, httpAuthScheme);
            }
        },
        httpAuthSchemes() {
            return _httpAuthSchemes;
        },
        setHttpAuthSchemeProvider(httpAuthSchemeProvider) {
            _httpAuthSchemeProvider = httpAuthSchemeProvider;
        },
        httpAuthSchemeProvider() {
            return _httpAuthSchemeProvider;
        },
        setCredentials(credentials) {
            _credentials = credentials;
        },
        credentials() {
            return _credentials;
        },
    };
};
const resolveHttpAuthRuntimeConfig = (config) => {
    return {
        httpAuthSchemes: config.httpAuthSchemes(),
        httpAuthSchemeProvider: config.httpAuthSchemeProvider(),
        credentials: config.credentials(),
    };
};

const resolveRuntimeExtensions = (runtimeConfig, extensions) => {
    const extensionConfiguration = Object.assign(regionConfigResolver.getAwsRegionExtensionConfiguration(runtimeConfig), smithyClient.getDefaultExtensionConfiguration(runtimeConfig), protocolHttp.getHttpHandlerExtensionConfiguration(runtimeConfig), getHttpAuthExtensionConfiguration(runtimeConfig));
    extensions.forEach((extension) => extension.configure(extensionConfiguration));
    return Object.assign(runtimeConfig, regionConfigResolver.resolveAwsRegionExtensionConfiguration(extensionConfiguration), smithyClient.resolveDefaultRuntimeConfig(extensionConfiguration), protocolHttp.resolveHttpHandlerRuntimeConfig(extensionConfiguration), resolveHttpAuthRuntimeConfig(extensionConfiguration));
};

class RekognitionStreamingClient extends smithyClient.Client {
    config;
    constructor(...[configuration]) {
        const _config_0 = runtimeConfig.getRuntimeConfig(configuration || {});
        super(_config_0);
        this.initConfig = _config_0;
        const _config_1 = resolveClientEndpointParameters(_config_0);
        const _config_2 = middlewareUserAgent.resolveUserAgentConfig(_config_1);
        const _config_3 = middlewareRetry.resolveRetryConfig(_config_2);
        const _config_4 = configResolver.resolveRegionConfig(_config_3);
        const _config_5 = middlewareHostHeader.resolveHostHeaderConfig(_config_4);
        const _config_6 = middlewareEndpoint.resolveEndpointConfig(_config_5);
        const _config_7 = eventstreamSerdeConfigResolver.resolveEventStreamSerdeConfig(_config_6);
        const _config_8 = httpAuthSchemeProvider.resolveHttpAuthSchemeConfig(_config_7);
        const _config_9 = middlewareEventstream.resolveEventStreamConfig(_config_8);
        const _config_10 = middlewareWebsocket.resolveWebSocketConfig(_config_9);
        const _config_11 = resolveRuntimeExtensions(_config_10, configuration?.extensions || []);
        this.config = _config_11;
        this.middlewareStack.use(schema.getSchemaSerdePlugin(this.config));
        this.middlewareStack.use(middlewareUserAgent.getUserAgentPlugin(this.config));
        this.middlewareStack.use(middlewareRetry.getRetryPlugin(this.config));
        this.middlewareStack.use(middlewareContentLength.getContentLengthPlugin(this.config));
        this.middlewareStack.use(middlewareHostHeader.getHostHeaderPlugin(this.config));
        this.middlewareStack.use(middlewareLogger.getLoggerPlugin(this.config));
        this.middlewareStack.use(middlewareRecursionDetection.getRecursionDetectionPlugin(this.config));
        this.middlewareStack.use(core.getHttpAuthSchemeEndpointRuleSetPlugin(this.config, {
            httpAuthSchemeParametersProvider: httpAuthSchemeProvider.defaultRekognitionStreamingHttpAuthSchemeParametersProvider,
            identityProviderConfigProvider: async (config) => new core.DefaultIdentityProviderConfig({
                "aws.auth#sigv4": config.credentials,
            }),
        }));
        this.middlewareStack.use(core.getHttpSigningPlugin(this.config));
    }
    destroy() {
        super.destroy();
    }
}

class RekognitionStreamingServiceException extends smithyClient.ServiceException {
    constructor(options) {
        super(options);
        Object.setPrototypeOf(this, RekognitionStreamingServiceException.prototype);
    }
}

class AccessDeniedException extends RekognitionStreamingServiceException {
    name = "AccessDeniedException";
    $fault = "client";
    Message;
    Code;
    constructor(opts) {
        super({
            name: "AccessDeniedException",
            $fault: "client",
            ...opts,
        });
        Object.setPrototypeOf(this, AccessDeniedException.prototype);
        this.Message = opts.Message;
        this.Code = opts.Code;
    }
}
class InternalServerException extends RekognitionStreamingServiceException {
    name = "InternalServerException";
    $fault = "server";
    Message;
    Code;
    constructor(opts) {
        super({
            name: "InternalServerException",
            $fault: "server",
            ...opts,
        });
        Object.setPrototypeOf(this, InternalServerException.prototype);
        this.Message = opts.Message;
        this.Code = opts.Code;
    }
}
class ServiceQuotaExceededException extends RekognitionStreamingServiceException {
    name = "ServiceQuotaExceededException";
    $fault = "client";
    Message;
    Code;
    constructor(opts) {
        super({
            name: "ServiceQuotaExceededException",
            $fault: "client",
            ...opts,
        });
        Object.setPrototypeOf(this, ServiceQuotaExceededException.prototype);
        this.Message = opts.Message;
        this.Code = opts.Code;
    }
}
class ServiceUnavailableException extends RekognitionStreamingServiceException {
    name = "ServiceUnavailableException";
    $fault = "server";
    Message;
    Code;
    constructor(opts) {
        super({
            name: "ServiceUnavailableException",
            $fault: "server",
            ...opts,
        });
        Object.setPrototypeOf(this, ServiceUnavailableException.prototype);
        this.Message = opts.Message;
        this.Code = opts.Code;
    }
}
class ThrottlingException extends RekognitionStreamingServiceException {
    name = "ThrottlingException";
    $fault = "client";
    Message;
    Code;
    constructor(opts) {
        super({
            name: "ThrottlingException",
            $fault: "client",
            ...opts,
        });
        Object.setPrototypeOf(this, ThrottlingException.prototype);
        this.Message = opts.Message;
        this.Code = opts.Code;
    }
}
class ValidationException extends RekognitionStreamingServiceException {
    name = "ValidationException";
    $fault = "client";
    Message;
    Code;
    constructor(opts) {
        super({
            name: "ValidationException",
            $fault: "client",
            ...opts,
        });
        Object.setPrototypeOf(this, ValidationException.prototype);
        this.Message = opts.Message;
        this.Code = opts.Code;
    }
}
class SessionNotFoundException extends RekognitionStreamingServiceException {
    name = "SessionNotFoundException";
    $fault = "client";
    Message;
    Code;
    constructor(opts) {
        super({
            name: "SessionNotFoundException",
            $fault: "client",
            ...opts,
        });
        Object.setPrototypeOf(this, SessionNotFoundException.prototype);
        this.Message = opts.Message;
        this.Code = opts.Code;
    }
}

const _ADE = "AccessDeniedException";
const _BB = "BoundingBox";
const _BFDT = "BlazeFaceDetectionThreshold";
const _C = "Code";
const _CC = "ChallengeConfig";
const _CCST = "CurrentColorStartTimestamp";
const _CCl = "ClientChallenge";
const _CCu = "CurrentColor";
const _CD = "ColorDisplayed";
const _CE = "ChallengeEvent";
const _CI = "ChallengeId";
const _CS = "ColorSequence";
const _CSIE = "ClientSessionInformationEvent";
const _CSo = "ColorSequences";
const _CV = "ChallengeVersions";
const _CX = "CenterX";
const _CY = "CenterY";
const _Ch = "Challenge";
const _DD = "DownscrollDuration";
const _DE = "DisconnectionEvent";
const _FC = "FreshnessColor";
const _FDD = "FlatDisplayDuration";
const _FDITPET = "FaceDetectedInTargetPositionEndTimestamp";
const _FDITPST = "FaceDetectedInTargetPositionStartTimestamp";
const _FDT = "FaceDistanceThreshold";
const _FDTM = "FaceDistanceThresholdMin";
const _FDTMa = "FaceDistanceThresholdMax";
const _FIHT = "FaceIouHeightThreshold";
const _FIWT = "FaceIouWidthThreshold";
const _FMALC = "FaceMovementAndLightChallenge";
const _FMALCC = "FaceMovementAndLightClientChallenge";
const _FMALSC = "FaceMovementAndLightServerChallenge";
const _FMC = "FaceMovementChallenge";
const _FMCC = "FaceMovementClientChallenge";
const _FMSC = "FaceMovementServerChallenge";
const _H = "Height";
const _IF = "InitialFace";
const _IFDT = "InitialFaceDetectedTimestamp";
const _ISE = "InternalServerException";
const _L = "Left";
const _LCT = "LightChallengeType";
const _LRS = "LivenessRequestStream";
const _LRSi = "LivenessResponseStream";
const _M = "Message";
const _OFT = "OvalFitTimeout";
const _OHWR = "OvalHeightWidthRatio";
const _OIHT = "OvalIouHeightThreshold";
const _OIT = "OvalIouThreshold";
const _OIWT = "OvalIouWidthThreshold";
const _OP = "OvalParameters";
const _PC = "PreviousColor";
const _RGB = "RGB";
const _SC = "ServerChallenge";
const _SFLS = "StartFaceLivenessSession";
const _SFLSR = "StartFaceLivenessSessionRequest";
const _SFLSRt = "StartFaceLivenessSessionResponse";
const _SI = "SessionInformation";
const _SIe = "SessionId";
const _SN = "SequenceNumber";
const _SNFE = "SessionNotFoundException";
const _SQEE = "ServiceQuotaExceededException";
const _SSIE = "ServerSessionInformationEvent";
const _SUE = "ServiceUnavailableException";
const _T = "Top";
const _TE = "ThrottlingException";
const _TF = "TargetFace";
const _TM = "TimestampMillis";
const _Ty = "Type";
const _V = "Version";
const _VC = "VideoChunk";
const _VE = "ValidationException";
const _VET = "VideoEndTimestamp";
const _VEi = "VideoEvent";
const _VH = "VideoHeight";
const _VST = "VideoStartTimestamp";
const _VW = "VideoWidth";
const _W = "Width";
const _c = "client";
const _e = "error";
const _h = "http";
const _hE = "httpError";
const _hH = "httpHeader";
const _s = "server";
const _sm = "smithy.ts.sdk.synthetic.com.amazonaws.rekognitionstreaming";
const _st = "streaming";
const _xarslcv = "x-amz-rekognition-streaming-liveness-challenge-versions";
const _xarslsi = "x-amz-rekognition-streaming-liveness-session-id";
const _xarslvh = "x-amz-rekognition-streaming-liveness-video-height";
const _xarslvw = "x-amz-rekognition-streaming-liveness-video-width";
const n0 = "com.amazonaws.rekognitionstreaming";
var AccessDeniedException$ = [-3, n0, _ADE,
    { [_e]: _c, [_hE]: 403 },
    [_M, _C],
    [0, 0]
];
schema.TypeRegistry.for(n0).registerError(AccessDeniedException$, AccessDeniedException);
var BoundingBox$ = [3, n0, _BB,
    0,
    [_W, _H, _L, _T],
    [1, 1, 1, 1]
];
var ChallengeConfig$ = [3, n0, _CC,
    0,
    [_BFDT, _FDTM, _FDT, _FDTMa, _OIT, _OHWR, _OIWT, _OIHT, _FIWT, _FIHT, _OFT],
    [1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1]
];
var ChallengeEvent$ = [3, n0, _CE,
    0,
    [_V, _Ty],
    [0, 0]
];
var ClientSessionInformationEvent$ = [3, n0, _CSIE,
    0,
    [_Ch],
    [() => ClientChallenge$]
];
var ColorDisplayed$ = [3, n0, _CD,
    0,
    [_CCu, _PC, _SN, _CCST],
    [() => FreshnessColor$, () => FreshnessColor$, 1, 1]
];
var ColorSequence$ = [3, n0, _CS,
    0,
    [_FC, _DD, _FDD],
    [() => FreshnessColor$, 1, 1]
];
var DisconnectionEvent$ = [3, n0, _DE,
    0,
    [_TM],
    [1]
];
var FaceMovementAndLightClientChallenge$ = [3, n0, _FMALCC,
    0,
    [_CI, _VST, _VET, _IF, _TF, _CD],
    [0, 1, 1, () => InitialFace$, () => TargetFace$, () => ColorDisplayed$]
];
var FaceMovementAndLightServerChallenge$ = [3, n0, _FMALSC,
    0,
    [_OP, _LCT, _CC, _CSo],
    [() => OvalParameters$, 0, () => ChallengeConfig$, () => ColorSequences]
];
var FaceMovementClientChallenge$ = [3, n0, _FMCC,
    0,
    [_CI, _VST, _VET, _IF, _TF],
    [0, 1, 1, () => InitialFace$, () => TargetFace$]
];
var FaceMovementServerChallenge$ = [3, n0, _FMSC,
    0,
    [_OP, _CC],
    [() => OvalParameters$, () => ChallengeConfig$]
];
var FreshnessColor$ = [3, n0, _FC,
    0,
    [_RGB],
    [64 | 1]
];
var InitialFace$ = [3, n0, _IF,
    0,
    [_BB, _IFDT],
    [() => BoundingBox$, 1]
];
var InternalServerException$ = [-3, n0, _ISE,
    { [_e]: _s, [_hE]: 500 },
    [_M, _C],
    [0, 0]
];
schema.TypeRegistry.for(n0).registerError(InternalServerException$, InternalServerException);
var OvalParameters$ = [3, n0, _OP,
    0,
    [_W, _H, _CX, _CY],
    [1, 1, 1, 1]
];
var ServerSessionInformationEvent$ = [3, n0, _SSIE,
    0,
    [_SI],
    [() => SessionInformation$]
];
var ServiceQuotaExceededException$ = [-3, n0, _SQEE,
    { [_e]: _c, [_hE]: 429 },
    [_M, _C],
    [0, 0]
];
schema.TypeRegistry.for(n0).registerError(ServiceQuotaExceededException$, ServiceQuotaExceededException);
var ServiceUnavailableException$ = [-3, n0, _SUE,
    { [_e]: _s, [_hE]: 503 },
    [_M, _C],
    [0, 0]
];
schema.TypeRegistry.for(n0).registerError(ServiceUnavailableException$, ServiceUnavailableException);
var SessionInformation$ = [3, n0, _SI,
    0,
    [_Ch],
    [() => ServerChallenge$]
];
var SessionNotFoundException$ = [-3, n0, _SNFE,
    { [_e]: _c, [_hE]: 400 },
    [_M, _C],
    [0, 0]
];
schema.TypeRegistry.for(n0).registerError(SessionNotFoundException$, SessionNotFoundException);
var StartFaceLivenessSessionRequest$ = [3, n0, _SFLSR,
    0,
    [_SIe, _VW, _VH, _CV, _LRS],
    [[0, { [_hH]: _xarslsi }], [0, { [_hH]: _xarslvw }], [0, { [_hH]: _xarslvh }], [0, { [_hH]: _xarslcv }], [() => LivenessRequestStream$, 16]]
];
var StartFaceLivenessSessionResponse$ = [3, n0, _SFLSRt,
    0,
    [_SIe, _LRSi],
    [[0, { [_hH]: _xarslsi }], [() => LivenessResponseStream$, 16]]
];
var TargetFace$ = [3, n0, _TF,
    0,
    [_BB, _FDITPST, _FDITPET],
    [() => BoundingBox$, 1, 1]
];
var ThrottlingException$ = [-3, n0, _TE,
    { [_e]: _c, [_hE]: 429 },
    [_M, _C],
    [0, 0]
];
schema.TypeRegistry.for(n0).registerError(ThrottlingException$, ThrottlingException);
var ValidationException$ = [-3, n0, _VE,
    { [_e]: _c, [_hE]: 400 },
    [_M, _C],
    [0, 0]
];
schema.TypeRegistry.for(n0).registerError(ValidationException$, ValidationException);
var VideoEvent$ = [3, n0, _VEi,
    0,
    [_VC, _TM],
    [21, 1]
];
var RekognitionStreamingServiceException$ = [-3, _sm, "RekognitionStreamingServiceException", 0, [], []];
schema.TypeRegistry.for(_sm).registerError(RekognitionStreamingServiceException$, RekognitionStreamingServiceException);
var ColorSequences = [1, n0, _CSo,
    0, () => ColorSequence$
];
var ClientChallenge$ = [4, n0, _CCl,
    0,
    [_FMALC, _FMC],
    [() => FaceMovementAndLightClientChallenge$, () => FaceMovementClientChallenge$]
];
var LivenessRequestStream$ = [4, n0, _LRS,
    { [_st]: 1 },
    [_VEi, _CSIE],
    [() => VideoEvent$, () => ClientSessionInformationEvent$]
];
var LivenessResponseStream$ = [4, n0, _LRSi,
    { [_st]: 1 },
    [_SSIE, _DE, _CE, _VE, _ISE, _TE, _SQEE, _SUE],
    [() => ServerSessionInformationEvent$, () => DisconnectionEvent$, () => ChallengeEvent$, [() => ValidationException$, 0], [() => InternalServerException$, 0], [() => ThrottlingException$, 0], [() => ServiceQuotaExceededException$, 0], [() => ServiceUnavailableException$, 0]]
];
var ServerChallenge$ = [4, n0, _SC,
    0,
    [_FMALC, _FMC],
    [() => FaceMovementAndLightServerChallenge$, () => FaceMovementServerChallenge$]
];
var StartFaceLivenessSession$ = [9, n0, _SFLS,
    { [_h]: ["POST", "/start-face-liveness-session", 200] }, () => StartFaceLivenessSessionRequest$, () => StartFaceLivenessSessionResponse$
];

class StartFaceLivenessSessionCommand extends smithyClient.Command
    .classBuilder()
    .ep(commonParams)
    .m(function (Command, cs, config, o) {
    return [
        middlewareEndpoint.getEndpointPlugin(config, Command.getEndpointParameterInstructions()),
        middlewareEventstream.getEventStreamPlugin(config),
        middlewareWebsocket.getWebSocketPlugin(config, {
            headerPrefix: 'x-amz-rekognition-streaming-liveness-',
        }),
    ];
})
    .s("RekognitionStreamingService", "StartFaceLivenessSession", {
    eventStream: {
        input: true,
        output: true,
    },
})
    .n("RekognitionStreamingClient", "StartFaceLivenessSessionCommand")
    .sc(StartFaceLivenessSession$)
    .build() {
}

const commands = {
    StartFaceLivenessSessionCommand,
};
class RekognitionStreaming extends RekognitionStreamingClient {
}
smithyClient.createAggregatedClient(commands, RekognitionStreaming);

const ChallengeType = {
    FACE_MOVEMENT_AND_LIGHT_CHALLENGE: "FaceMovementAndLightChallenge",
    FACE_MOVEMENT_CHALLENGE: "FaceMovementChallenge",
};
const LightChallengeType = {
    SEQUENTIAL: "SEQUENTIAL",
};

Object.defineProperty(exports, "$Command", {
    enumerable: true,
    get: function () { return smithyClient.Command; }
});
Object.defineProperty(exports, "__Client", {
    enumerable: true,
    get: function () { return smithyClient.Client; }
});
exports.AccessDeniedException = AccessDeniedException;
exports.AccessDeniedException$ = AccessDeniedException$;
exports.BoundingBox$ = BoundingBox$;
exports.ChallengeConfig$ = ChallengeConfig$;
exports.ChallengeEvent$ = ChallengeEvent$;
exports.ChallengeType = ChallengeType;
exports.ClientChallenge$ = ClientChallenge$;
exports.ClientSessionInformationEvent$ = ClientSessionInformationEvent$;
exports.ColorDisplayed$ = ColorDisplayed$;
exports.ColorSequence$ = ColorSequence$;
exports.DisconnectionEvent$ = DisconnectionEvent$;
exports.FaceMovementAndLightClientChallenge$ = FaceMovementAndLightClientChallenge$;
exports.FaceMovementAndLightServerChallenge$ = FaceMovementAndLightServerChallenge$;
exports.FaceMovementClientChallenge$ = FaceMovementClientChallenge$;
exports.FaceMovementServerChallenge$ = FaceMovementServerChallenge$;
exports.FreshnessColor$ = FreshnessColor$;
exports.InitialFace$ = InitialFace$;
exports.InternalServerException = InternalServerException;
exports.InternalServerException$ = InternalServerException$;
exports.LightChallengeType = LightChallengeType;
exports.LivenessRequestStream$ = LivenessRequestStream$;
exports.LivenessResponseStream$ = LivenessResponseStream$;
exports.OvalParameters$ = OvalParameters$;
exports.RekognitionStreaming = RekognitionStreaming;
exports.RekognitionStreamingClient = RekognitionStreamingClient;
exports.RekognitionStreamingServiceException = RekognitionStreamingServiceException;
exports.RekognitionStreamingServiceException$ = RekognitionStreamingServiceException$;
exports.ServerChallenge$ = ServerChallenge$;
exports.ServerSessionInformationEvent$ = ServerSessionInformationEvent$;
exports.ServiceQuotaExceededException = ServiceQuotaExceededException;
exports.ServiceQuotaExceededException$ = ServiceQuotaExceededException$;
exports.ServiceUnavailableException = ServiceUnavailableException;
exports.ServiceUnavailableException$ = ServiceUnavailableException$;
exports.SessionInformation$ = SessionInformation$;
exports.SessionNotFoundException = SessionNotFoundException;
exports.SessionNotFoundException$ = SessionNotFoundException$;
exports.StartFaceLivenessSession$ = StartFaceLivenessSession$;
exports.StartFaceLivenessSessionCommand = StartFaceLivenessSessionCommand;
exports.StartFaceLivenessSessionRequest$ = StartFaceLivenessSessionRequest$;
exports.StartFaceLivenessSessionResponse$ = StartFaceLivenessSessionResponse$;
exports.TargetFace$ = TargetFace$;
exports.ThrottlingException = ThrottlingException;
exports.ThrottlingException$ = ThrottlingException$;
exports.ValidationException = ValidationException;
exports.ValidationException$ = ValidationException$;
exports.VideoEvent$ = VideoEvent$;
