// @aws-amplify/ui-react-liveness/dist/esm/components/FaceLivenessDetector/service/utils/ColorSequenceDisplay/ColorSequenceDisplay.mjs
import { __classPrivateFieldSet, __classPrivateFieldGet } from 'tslib';
import { isFunction } from '@aws-amplify/ui';
import { TICK_RATE } from '../constants.mjs';

var _ColorSequenceDisplay_instances, _ColorSequenceDisplay_sequence, _ColorSequenceDisplay_previousSequence, _ColorSequenceDisplay_colorStage, _ColorSequenceDisplay_sequenceIndex, _ColorSequenceDisplay_colorSequences, _ColorSequenceDisplay_isFirstTick, _ColorSequenceDisplay_lastColorStageChangeTimestamp, _ColorSequenceDisplay_isFlatStage, _ColorSequenceDisplay_isScrollingStage, _ColorSequenceDisplay_startColorSequence, _ColorSequenceDisplay_handleSequenceChange;
var ColorStageType;
(function (ColorStageType) {
    ColorStageType[ColorStageType["Scrolling"] = 0] = "Scrolling";
    ColorStageType[ColorStageType["Flat"] = 1] = "Flat";
})(ColorStageType || (ColorStageType = {}));
class ColorSequenceDisplay {
    /**
     * Iterates over provided color sequences and executes sequence event callbacks
     *
     * @param {ColorSequences} colorSequences array of color sequences to iterate over
     */
    constructor(colorSequences) {
        _ColorSequenceDisplay_instances.add(this);
        /**
         * the current color sequence used for flat display and the prev color when scrolling
         */
        _ColorSequenceDisplay_sequence.set(this, void 0);
        /**
         * previous color sequence, during flat display curr === prev and during scroll it is the prev indexed color
         */
        _ColorSequenceDisplay_previousSequence.set(this, void 0);
        /**
         * current ColorStage, initialize to 'FLAT'
         */
        _ColorSequenceDisplay_colorStage.set(this, ColorStageType.Flat);
        /**
         * current color sequence index (black flat, red scrolling, etc)
         */
        _ColorSequenceDisplay_sequenceIndex.set(this, 0);
        _ColorSequenceDisplay_colorSequences.set(this, void 0);
        _ColorSequenceDisplay_isFirstTick.set(this, true);
        _ColorSequenceDisplay_lastColorStageChangeTimestamp.set(this, 0);
        __classPrivateFieldSet(this, _ColorSequenceDisplay_colorSequences, colorSequences, "f");
        __classPrivateFieldSet(this, _ColorSequenceDisplay_sequence, colorSequences[0], "f");
        __classPrivateFieldSet(this, _ColorSequenceDisplay_previousSequence, colorSequences[0], "f");
    }
    /**
     * Start sequence iteration and execute event callbacks
     *
     * @async
     * @param {StartSequencesParams} params Sequence event handlers
     * @returns {Promise<boolean>} Resolves to true when complete
     */
    async startSequences(params) {
        return new Promise((resolve) => {
            setTimeout(() => {
                __classPrivateFieldGet(this, _ColorSequenceDisplay_instances, "m", _ColorSequenceDisplay_startColorSequence).call(this, { ...params, resolve });
            }, Math.min(TICK_RATE));
        });
    }
}
_ColorSequenceDisplay_sequence = new WeakMap(), _ColorSequenceDisplay_previousSequence = new WeakMap(), _ColorSequenceDisplay_colorStage = new WeakMap(), _ColorSequenceDisplay_sequenceIndex = new WeakMap(), _ColorSequenceDisplay_colorSequences = new WeakMap(), _ColorSequenceDisplay_isFirstTick = new WeakMap(), _ColorSequenceDisplay_lastColorStageChangeTimestamp = new WeakMap(), _ColorSequenceDisplay_instances = new WeakSet(), _ColorSequenceDisplay_isFlatStage = function _ColorSequenceDisplay_isFlatStage() {
    return __classPrivateFieldGet(this, _ColorSequenceDisplay_colorStage, "f") === ColorStageType.Flat;
}, _ColorSequenceDisplay_isScrollingStage = function _ColorSequenceDisplay_isScrollingStage() {
    return __classPrivateFieldGet(this, _ColorSequenceDisplay_colorStage, "f") === ColorStageType.Scrolling;
}, _ColorSequenceDisplay_startColorSequence = function _ColorSequenceDisplay_startColorSequence({ onSequenceChange, onSequenceColorChange, onSequenceStart, onSequencesComplete, resolve, }) {
    if (isFunction(onSequenceStart)) {
        onSequenceStart();
    }
    const sequenceStartTime = Date.now();
    let timeSinceLastColorStageChange = sequenceStartTime - __classPrivateFieldGet(this, _ColorSequenceDisplay_lastColorStageChangeTimestamp, "f");
    // Send a colorStart time only for the first tick of the first color
    if (__classPrivateFieldGet(this, _ColorSequenceDisplay_isFirstTick, "f")) {
        __classPrivateFieldSet(this, _ColorSequenceDisplay_lastColorStageChangeTimestamp, Date.now(), "f");
        __classPrivateFieldSet(this, _ColorSequenceDisplay_isFirstTick, false, "f");
        // initial sequence change
        if (isFunction(onSequenceChange)) {
            onSequenceChange({
                prevSequenceColor: __classPrivateFieldGet(this, _ColorSequenceDisplay_previousSequence, "f").color,
                sequenceColor: __classPrivateFieldGet(this, _ColorSequenceDisplay_sequence, "f").color,
                sequenceIndex: __classPrivateFieldGet(this, _ColorSequenceDisplay_sequenceIndex, "f"),
                sequenceStartTime,
            });
        }
    }
    // Every 10 ms tick we will check if the threshold for flat or scrolling, if so we will try to go to the next stage
    if ((__classPrivateFieldGet(this, _ColorSequenceDisplay_instances, "m", _ColorSequenceDisplay_isFlatStage).call(this) &&
        timeSinceLastColorStageChange >= __classPrivateFieldGet(this, _ColorSequenceDisplay_sequence, "f").flatDisplayDuration) ||
        (__classPrivateFieldGet(this, _ColorSequenceDisplay_instances, "m", _ColorSequenceDisplay_isScrollingStage).call(this) &&
            timeSinceLastColorStageChange >= __classPrivateFieldGet(this, _ColorSequenceDisplay_sequence, "f").downscrollDuration)) {
        __classPrivateFieldGet(this, _ColorSequenceDisplay_instances, "m", _ColorSequenceDisplay_handleSequenceChange).call(this, { sequenceStartTime, onSequenceChange });
        timeSinceLastColorStageChange = 0;
    }
    const hasRemainingSequences = __classPrivateFieldGet(this, _ColorSequenceDisplay_sequenceIndex, "f") < __classPrivateFieldGet(this, _ColorSequenceDisplay_colorSequences, "f").length;
    // Every 10 ms tick we will update the colors displayed
    if (hasRemainingSequences) {
        const heightFraction = timeSinceLastColorStageChange /
            (__classPrivateFieldGet(this, _ColorSequenceDisplay_instances, "m", _ColorSequenceDisplay_isScrollingStage).call(this)
                ? __classPrivateFieldGet(this, _ColorSequenceDisplay_sequence, "f").downscrollDuration
                : __classPrivateFieldGet(this, _ColorSequenceDisplay_sequence, "f").flatDisplayDuration);
        if (isFunction(onSequenceColorChange)) {
            onSequenceColorChange({
                sequenceColor: __classPrivateFieldGet(this, _ColorSequenceDisplay_sequence, "f").color,
                heightFraction,
                prevSequenceColor: __classPrivateFieldGet(this, _ColorSequenceDisplay_previousSequence, "f").color,
            });
        }
        resolve(false);
    }
    else {
        if (isFunction(onSequencesComplete)) {
            onSequencesComplete();
        }
        resolve(true);
    }
}, _ColorSequenceDisplay_handleSequenceChange = function _ColorSequenceDisplay_handleSequenceChange({ sequenceStartTime, onSequenceChange, }) {
    __classPrivateFieldSet(this, _ColorSequenceDisplay_previousSequence, __classPrivateFieldGet(this, _ColorSequenceDisplay_sequence, "f"), "f");
    if (__classPrivateFieldGet(this, _ColorSequenceDisplay_instances, "m", _ColorSequenceDisplay_isFlatStage).call(this)) {
        __classPrivateFieldSet(this, _ColorSequenceDisplay_sequenceIndex, __classPrivateFieldGet(this, _ColorSequenceDisplay_sequenceIndex, "f") + 1, "f");
        __classPrivateFieldSet(this, _ColorSequenceDisplay_colorStage, ColorStageType.Scrolling, "f");
    }
    else if (__classPrivateFieldGet(this, _ColorSequenceDisplay_instances, "m", _ColorSequenceDisplay_isScrollingStage).call(this)) {
        const nextColorSequence = __classPrivateFieldGet(this, _ColorSequenceDisplay_colorSequences, "f")[__classPrivateFieldGet(this, _ColorSequenceDisplay_sequenceIndex, "f")];
        if (nextColorSequence.flatDisplayDuration > 0) {
            __classPrivateFieldSet(this, _ColorSequenceDisplay_colorStage, ColorStageType.Flat, "f");
        }
        else {
            __classPrivateFieldSet(this, _ColorSequenceDisplay_sequenceIndex, __classPrivateFieldGet(this, _ColorSequenceDisplay_sequenceIndex, "f") + 1, "f");
        }
    }
    __classPrivateFieldSet(this, _ColorSequenceDisplay_sequence, __classPrivateFieldGet(this, _ColorSequenceDisplay_colorSequences, "f")[__classPrivateFieldGet(this, _ColorSequenceDisplay_sequenceIndex, "f")], "f");
    __classPrivateFieldSet(this, _ColorSequenceDisplay_lastColorStageChangeTimestamp, Date.now(), "f");
    if (__classPrivateFieldGet(this, _ColorSequenceDisplay_sequence, "f")) {
        if (isFunction(onSequenceChange)) {
            onSequenceChange({
                prevSequenceColor: __classPrivateFieldGet(this, _ColorSequenceDisplay_previousSequence, "f").color,
                sequenceColor: __classPrivateFieldGet(this, _ColorSequenceDisplay_sequence, "f").color,
                sequenceIndex: __classPrivateFieldGet(this, _ColorSequenceDisplay_sequenceIndex, "f"),
                sequenceStartTime: sequenceStartTime,
            });
        }
    }
};

export { ColorSequenceDisplay };
