// @aws-amplify/ui-react-liveness/dist/esm/components/FaceLivenessDetector/service/utils/constants.mjs
// Face distance is calculated as pupilDistance / ovalWidth.
// The further away you are from the camera the distance between your pupils will decrease, thus lowering the threshold values.
// Constants from science team to determine ocular distance (space between eyes)
const PUPIL_DISTANCE_WEIGHT = 2.0;
const FACE_HEIGHT_WEIGHT = 1.8;
// Constants from science team to find face match percentage
const FACE_MATCH_RANGE_MIN = 0;
const FACE_MATCH_RANGE_MAX = 1;
const FACE_MATCH_WEIGHT_MIN = 0.25;
const FACE_MATCH_WEIGHT_MAX = 0.75;
const OVAL_HEIGHT_WIDTH_RATIO = 1.618;
const WS_CLOSURE_CODE = {
    SUCCESS_CODE: 1000,
    DEFAULT_ERROR_CODE: 4000,
    FACE_FIT_TIMEOUT: 4001,
    USER_CANCEL: 4003,
    RUNTIME_ERROR: 4005,
    USER_ERROR_DURING_CONNECTION: 4007,
};
// number in milliseconds to record into each video chunk.
// see https://developer.mozilla.org/en-US/docs/Web/API/MediaRecorder/start#timeslice
const TIME_SLICE = 1000;
// in MS, the rate at which colors are rendered/checked
const TICK_RATE = 10;
/**
 * The number of seconds before the presigned URL expires.
 * Used to override aws sdk default value of 60
 */
const REQUEST_EXPIRY = 299;
/**
 * The maximum time in milliseconds that the connection phase of a request
 * may take before the connection attempt is abandoned.
 */
const CONNECTION_TIMEOUT = 10000;
const FACE_MOVEMENT_AND_LIGHT_CHALLENGE = {
    type: 'FaceMovementAndLightChallenge',
    version: '2.0.0',
};
const FACE_MOVEMENT_CHALLENGE = {
    type: 'FaceMovementChallenge',
    version: '1.0.0',
};
const SUPPORTED_CHALLENGES = [
    FACE_MOVEMENT_AND_LIGHT_CHALLENGE,
    FACE_MOVEMENT_CHALLENGE,
];
const queryParameterString = SUPPORTED_CHALLENGES.map((challenge) => `${challenge.type}_${challenge.version}`).join(',');

export { CONNECTION_TIMEOUT, FACE_HEIGHT_WEIGHT, FACE_MATCH_RANGE_MAX, FACE_MATCH_RANGE_MIN, FACE_MATCH_WEIGHT_MAX, FACE_MATCH_WEIGHT_MIN, FACE_MOVEMENT_AND_LIGHT_CHALLENGE, FACE_MOVEMENT_CHALLENGE, OVAL_HEIGHT_WIDTH_RATIO, PUPIL_DISTANCE_WEIGHT, REQUEST_EXPIRY, SUPPORTED_CHALLENGES, TICK_RATE, TIME_SLICE, WS_CLOSURE_CODE, queryParameterString };
