// @aws-amplify/ui-react-liveness/dist/esm/components/FaceLivenessDetector/service/utils/StreamRecorder/StreamRecorder.mjs
import { __classPrivateFieldSet, __classPrivateFieldGet } from 'tslib';
import { isUndefined } from '@aws-amplify/ui';
import { TIME_SLICE } from '../constants.mjs';

var _StreamRecorder_instances, _StreamRecorder_chunks, _StreamRecorder_recorder, _StreamRecorder_initialRecorder, _StreamRecorder_recordingStarted, _StreamRecorder_firstChunkTimestamp, _StreamRecorder_recorderEndTimestamp, _StreamRecorder_recorderStartTimestamp, _StreamRecorder_recordingStartTimestamp, _StreamRecorder_recorderStopped, _StreamRecorder_videoStream, _StreamRecorder_eventListeners, _StreamRecorder_clearRecordedChunks, _StreamRecorder_createReadableStream, _StreamRecorder_attachHandlers, _StreamRecorder_setupCallbacks, _StreamRecorder_cleanUpEventListeners;
class StreamRecorder {
    constructor(stream) {
        _StreamRecorder_instances.add(this);
        _StreamRecorder_chunks.set(this, void 0);
        _StreamRecorder_recorder.set(this, void 0);
        _StreamRecorder_initialRecorder.set(this, void 0);
        _StreamRecorder_recordingStarted.set(this, false);
        _StreamRecorder_firstChunkTimestamp.set(this, void 0);
        _StreamRecorder_recorderEndTimestamp.set(this, void 0);
        _StreamRecorder_recorderStartTimestamp.set(this, void 0);
        _StreamRecorder_recordingStartTimestamp.set(this, void 0);
        _StreamRecorder_recorderStopped.set(this, void 0);
        _StreamRecorder_videoStream.set(this, void 0);
        _StreamRecorder_eventListeners.set(this, void 0);
        if (typeof MediaRecorder === 'undefined') {
            throw Error('MediaRecorder is not supported by this browser');
        }
        __classPrivateFieldSet(this, _StreamRecorder_chunks, [], "f");
        __classPrivateFieldSet(this, _StreamRecorder_recorder, new MediaRecorder(stream, { bitsPerSecond: 1000000 }), "f");
        __classPrivateFieldSet(this, _StreamRecorder_initialRecorder, __classPrivateFieldGet(this, _StreamRecorder_recorder, "f"), "f");
        __classPrivateFieldSet(this, _StreamRecorder_videoStream, __classPrivateFieldGet(this, _StreamRecorder_instances, "m", _StreamRecorder_createReadableStream).call(this), "f");
        __classPrivateFieldSet(this, _StreamRecorder_eventListeners, {}, "f");
    }
    getVideoStream() {
        return __classPrivateFieldGet(this, _StreamRecorder_videoStream, "f");
    }
    setNewVideoStream(stream) {
        __classPrivateFieldGet(this, _StreamRecorder_instances, "m", _StreamRecorder_cleanUpEventListeners).call(this);
        __classPrivateFieldSet(this, _StreamRecorder_recorder, new MediaRecorder(stream, { bitsPerSecond: 1000000 }), "f");
        __classPrivateFieldGet(this, _StreamRecorder_instances, "m", _StreamRecorder_attachHandlers).call(this, __classPrivateFieldGet(this, _StreamRecorder_recorder, "f"));
    }
    dispatchStreamEvent(event) {
        const { type } = event;
        const data = type === 'streamStop' ? undefined : event.data;
        __classPrivateFieldGet(this, _StreamRecorder_recorder, "f").dispatchEvent(new MessageEvent(type, { data }));
    }
    getRecordingStartTimestamp() {
        if (isUndefined(__classPrivateFieldGet(this, _StreamRecorder_recorderStartTimestamp, "f")) ||
            isUndefined(__classPrivateFieldGet(this, _StreamRecorder_recordingStartTimestamp, "f"))) {
            throw new Error('Recording has not started');
        }
        /**
         * This calculation is provided by Science team after doing analysis
         * of unreliable .onstart() (this.#recorderStartTimestamp) timestamp that is
         * returned from mediaRecorder.
         */
        return Math.round(0.73 * (__classPrivateFieldGet(this, _StreamRecorder_recorderStartTimestamp, "f") - __classPrivateFieldGet(this, _StreamRecorder_recordingStartTimestamp, "f")) +
            __classPrivateFieldGet(this, _StreamRecorder_recordingStartTimestamp, "f"));
    }
    getRecordingEndedTimestamp() {
        if (isUndefined(__classPrivateFieldGet(this, _StreamRecorder_recorderEndTimestamp, "f"))) {
            throw new Error('Recording has not ended');
        }
        return __classPrivateFieldGet(this, _StreamRecorder_recorderEndTimestamp, "f");
    }
    startRecording() {
        __classPrivateFieldGet(this, _StreamRecorder_instances, "m", _StreamRecorder_clearRecordedChunks).call(this);
        __classPrivateFieldSet(this, _StreamRecorder_recordingStartTimestamp, Date.now(), "f");
        __classPrivateFieldGet(this, _StreamRecorder_recorder, "f").start(TIME_SLICE);
    }
    isRecording() {
        return __classPrivateFieldGet(this, _StreamRecorder_recorder, "f").state === 'recording';
    }
    getChunksLength() {
        return __classPrivateFieldGet(this, _StreamRecorder_chunks, "f").length;
    }
    hasRecordingStarted() {
        return __classPrivateFieldGet(this, _StreamRecorder_recordingStarted, "f") && __classPrivateFieldGet(this, _StreamRecorder_firstChunkTimestamp, "f") !== undefined;
    }
    async stopRecording() {
        if (this.isRecording()) {
            __classPrivateFieldGet(this, _StreamRecorder_recorder, "f").stop();
        }
        return __classPrivateFieldGet(this, _StreamRecorder_recorderStopped, "f");
    }
}
_StreamRecorder_chunks = new WeakMap(), _StreamRecorder_recorder = new WeakMap(), _StreamRecorder_initialRecorder = new WeakMap(), _StreamRecorder_recordingStarted = new WeakMap(), _StreamRecorder_firstChunkTimestamp = new WeakMap(), _StreamRecorder_recorderEndTimestamp = new WeakMap(), _StreamRecorder_recorderStartTimestamp = new WeakMap(), _StreamRecorder_recordingStartTimestamp = new WeakMap(), _StreamRecorder_recorderStopped = new WeakMap(), _StreamRecorder_videoStream = new WeakMap(), _StreamRecorder_eventListeners = new WeakMap(), _StreamRecorder_instances = new WeakSet(), _StreamRecorder_clearRecordedChunks = function _StreamRecorder_clearRecordedChunks() {
    __classPrivateFieldSet(this, _StreamRecorder_chunks, [], "f");
}, _StreamRecorder_createReadableStream = function _StreamRecorder_createReadableStream() {
    return new ReadableStream({
        start: (controller) => {
            __classPrivateFieldGet(this, _StreamRecorder_instances, "m", _StreamRecorder_attachHandlers).call(this, __classPrivateFieldGet(this, _StreamRecorder_recorder, "f"), controller);
        },
    });
}, _StreamRecorder_attachHandlers = function _StreamRecorder_attachHandlers(recorder, controller) {
    const onDataAvailableHandler = controller
        ? ({ data }) => {
            if (data && data.size > 0) {
                if (__classPrivateFieldGet(this, _StreamRecorder_chunks, "f").length === 0) {
                    __classPrivateFieldSet(this, _StreamRecorder_firstChunkTimestamp, Date.now(), "f");
                }
                __classPrivateFieldGet(this, _StreamRecorder_chunks, "f").push(data);
                controller.enqueue({ type: 'streamVideo', data });
            }
        }
        : ({ data }) => {
            __classPrivateFieldGet(this, _StreamRecorder_initialRecorder, "f").dispatchEvent(new MessageEvent('dataavailable', { data }));
        };
    recorder.ondataavailable = onDataAvailableHandler;
    const onSessionInfoHandler = controller
        ? (e) => {
            const { data } = e;
            controller.enqueue({ type: 'sessionInfo', data });
        }
        : (e) => {
            const { data } = e;
            __classPrivateFieldGet(this, _StreamRecorder_initialRecorder, "f").dispatchEvent(new MessageEvent('sessionInfo', { data }));
        };
    recorder.addEventListener('sessionInfo', onSessionInfoHandler);
    const onStreamStopHandler = controller
        ? () => {
            controller.enqueue({ type: 'streamStop' });
        }
        : () => {
            __classPrivateFieldGet(this, _StreamRecorder_initialRecorder, "f").dispatchEvent(new MessageEvent('streamStop'));
        };
    recorder.addEventListener('streamStop', onStreamStopHandler);
    const onCloseCodeHandler = controller
        ? (e) => {
            const { data } = e;
            controller.enqueue({ type: 'closeCode', data });
        }
        : (e) => {
            const { data } = e;
            __classPrivateFieldGet(this, _StreamRecorder_initialRecorder, "f").dispatchEvent(new MessageEvent('closeCode', { data }));
        };
    recorder.addEventListener('closeCode', onCloseCodeHandler);
    const onEndStreamHandler = controller
        ? () => {
            controller.close();
        }
        : () => {
            __classPrivateFieldGet(this, _StreamRecorder_initialRecorder, "f").dispatchEvent(new MessageEvent('endStream'));
        };
    recorder.addEventListener('endStream', onEndStreamHandler);
    __classPrivateFieldGet(this, _StreamRecorder_instances, "m", _StreamRecorder_setupCallbacks).call(this);
    __classPrivateFieldSet(this, _StreamRecorder_eventListeners, {
        endStream: onEndStreamHandler,
        closeCode: onCloseCodeHandler,
        streamStop: onStreamStopHandler,
        sessionInfo: onSessionInfoHandler,
        dataavailable: onDataAvailableHandler,
    }, "f");
}, _StreamRecorder_setupCallbacks = function _StreamRecorder_setupCallbacks() {
    __classPrivateFieldGet(this, _StreamRecorder_recorder, "f").onstart = () => {
        __classPrivateFieldSet(this, _StreamRecorder_recordingStarted, true, "f");
        __classPrivateFieldSet(this, _StreamRecorder_recorderStartTimestamp, Date.now(), "f");
    };
    __classPrivateFieldSet(this, _StreamRecorder_recorderStopped, new Promise((resolve) => {
        __classPrivateFieldGet(this, _StreamRecorder_recorder, "f").onstop = () => {
            __classPrivateFieldSet(this, _StreamRecorder_recorderEndTimestamp, Date.now(), "f");
            resolve();
        };
    }), "f");
    __classPrivateFieldGet(this, _StreamRecorder_recorder, "f").onerror = () => {
        this.stopRecording();
    };
}, _StreamRecorder_cleanUpEventListeners = function _StreamRecorder_cleanUpEventListeners() {
    const eventNames = Object.keys(__classPrivateFieldGet(this, _StreamRecorder_eventListeners, "f"));
    eventNames.forEach((name) => {
        __classPrivateFieldGet(this, _StreamRecorder_recorder, "f").removeEventListener(name, __classPrivateFieldGet(this, _StreamRecorder_eventListeners, "f")[name]);
    });
    __classPrivateFieldSet(this, _StreamRecorder_eventListeners, {}, "f");
};

export { StreamRecorder };
